### ISSUE TITLE
TwitchHttpClient can have a more accurate message

### ISSUE THREAD

**Issue by JDR-Ninja** [TwitchLib/TwitchLib.Api#210]
Hello,
The method TwitchLib.Api.Core.HttpCallHandlers.TwitchHttpClient.HandleWebException throw a BadTokenException when the api return 403.
But in some case, the returned message from the api is :
{
    "error": "Forbidden",
    "status": 403,
    "message": "channel points are not available for the broadcaster"
}
But TwitchHttpClient will hide this information and return a bad token message.
I tested this issue with my account (new twitch account) and called "Helix.ChannelPoints.GetCustomReward".
I searched for a long time before realizing that my token was correct, but the problem was that my channel did not support the functionality.

**Comment by Syzuna** [comment]
Yes we know this is not ideal but the library pre dates Helix and its standard error messages and back then we only had Status codes to work with.
Changing it would be a big change and most likely also would cause a lot of hacked together code.. so I guess this wont be done until v5 is removed and we got time to rewrite the library completely (sth we are planning for some time now).
Sorry :/

**Comment by bilago** [comment]
> Yes we know this is not ideal but the library pre dates Helix and its standard error messages and back then we only had Status codes to work with. Changing it would be a big change and most likely also would cause a lot of hacked together code.. so I guess this wont be done until v5 is removed and we got time to rewrite the library completely (sth we are planning for some time now). Sorry :/

Couldn't this be simply added as such, without breaking anything? I wouldn't see the harm in  adding the errorResp.ReasonPhrase if it's not null as part of the Exception Message.


    private void HandleWebException(HttpResponseMessage errorResp)
        {
            var moreInfo = errorResp.ReasonPhrase ?? string.Empty;
            switch (errorResp.StatusCode)
            {
                case HttpStatusCode.BadRequest:
                    throw new BadRequestException("Your request failed because either: \n 1. Your ClientID was invalid/not set. \n 2. Your refresh token was invalid. \n 3. You requested a username when the server was expecting a user ID."+ moreInfo);
                case HttpStatusCode.Unauthorized:
                    var authenticateHeader = errorResp.Headers.WwwAuthenticate;
                    if (authenticateHeader == null || authenticateHeader.Count <= 0)
                        throw new BadScopeException("Your request was blocked due to bad credentials (Do you have the right scope for your access token?)."+moreInfo);
                    else
                        throw new TokenExpiredException("Your request was blocked due to an expired Token. Please refresh your token and update your API instance settings." + moreInfo);
                case HttpStatusCode.NotFound:
                    throw new BadResourceException("The resource you tried to access was not valid." + moreInfo);
                case (HttpStatusCode)422:
                    throw new NotPartneredException("The resource you requested is only available to channels that have been partnered by Twitch." + moreInfo);
                case (HttpStatusCode)429:
                    errorResp.Headers.TryGetValues("Ratelimit-Reset", out var resetTime);
                    throw new TooManyRequestsException("You have reached your rate limit. Too many requests were made" + moreInfo, resetTime.FirstOrDefault());
                case HttpStatusCode.BadGateway:
                    throw new BadGatewayException("The API answered with a 502 Bad Gateway. Please retry your request" + moreInfo);
                case HttpStatusCode.GatewayTimeout:
                    throw new GatewayTimeoutException("The API answered with a 504 Gateway Timeout. Please retry your request" + moreInfo);
                case HttpStatusCode.InternalServerError:
                    throw new InternalServerErrorException("The API answered with a 500 Internal Server Error. Please retry your request" + moreInfo);
                case HttpStatusCode.Forbidden:
                    throw new BadTokenException("The token provided in the request did not match the associated user. Make sure the token you're using is from the resource owner (streamer? viewer?)" + moreInfo);
                default:
                    throw new HttpRequestException("Something went wrong during the request! Please try again later" + moreInfo);
            }
        }

**Comment by Syzuna** [comment]
While it isnt ideal it is at least sth.
With 3.9.0 you can now do:

```
            try
            {
               // API call
            }
            catch (HttpResponseException e)
            {
                Console.WriteLine(await e.HttpResponse.Content.ReadAsStringAsync());
            }
```

**Comment by Felk** [comment]
With a bunch of chat functionality being moved over to Helix, I suspect this will trip quite a few more people now. For example, when whispering a user Twitch might return a status code 403 with the message

> The recipient's settings prevent this sender from whispering them

but TwitchLib outputs this error, which can misguide people quite badly:

> TwitchLib.Api.Core.Exceptions.BadTokenException: The token provided in the request did not match the associated user. Make sure the token you're using is from the resource owner (streamer? viewer?)

Would it be acceptable to generify the error message a bit? I believe a less helpful message would be better than a misleading one.
